src/rewrite.ts

```diff
@@ -0,0 +1,54 @@
+export class RewriteRule {
+  find: RegExp;
+  replace: string;
+
+  constructor(find: RegExp, replace: string) {
+    this.find = find;
+    this.replace = replace;
+  }
+
+  apply(value: string): string {
+    return value.replace(this.find, this.replace);
+  }
+}
+
+export class RewriteRules {
+  private rules: RewriteRule[];
+
+  constructor() {
+    this.rules = [];
+  }
+
+  appendRule(rule: RewriteRule): RewriteRules {
+    this.rules.push(rule);
+    return this;
+  }
+
+  apply<T>(value: T): T {
+    if (typeof value === "object" && value !== null) {
+      // If the object is an array, iterate through each element and call the function recursively
+      if (Array.isArray(value)) {
+        return (value.map((v) => this.apply(v)) as any) as T;
+      }
+
+      // If the object is not an array, create a new object with the same keys,
+      // and call the function recursively on each value
+      const oldObj = value as { [key: string]: any };
+      const newObj: { [key: string]: any } = {};
+      for (const key of Object.keys(oldObj)) {
+        const newKey = this.apply(key);
+        const newValue = this.apply(oldObj[key]);
+        newObj[newKey] = newValue;
+      }
+      return newObj as T;
+    } else if (typeof value === "string") {
+      let s = value as string;
+      for (const rule of this.rules) {
+        s = rule.apply(value);
+      }
+      return (s as any) as T;
```

Comment:
**Question** Why is this first converted to any to then immediately be casted to T?

Reply:
The type of `s` was resolved as a string with our manual check `typeof value === "string"` and subsequent cast `let s = value as string`. However the top level recursed function has a generic type T which it needs to return. `string` and `T` don't overlap and the cast to `any` is required to allow us to cast it to T afterwards

Reply:
This is a limitation of TypeScript. Line 54 casts the generic type `T` to `string`. We need to cast `string` back to `T` in order to return it. The compiler does not allow casting `string` directly to `T` as there's no overlap of the types. Going via `any` is the way you do a "cast to anything" cast in TypeScript.